File: android-app/app/src/main/java/com/codex/quota/android/domain/NotificationPolicy.kt
package com.codex.quota.android.domain

enum class TaskState {
  Running,
  NeedsAuthorization,
  WaitingForReview,
}

enum class NotificationMode {
  Never,
  WhenChatGptUnfocused,
  Always,
}

enum class PhoneUrgency {
  Silent,
  Vibrate,
}

enum class BandAlertBehavior {
  SystemControlled,
}

data class AlertContext(
  val chatGptFocused: Boolean,
  val androidForeground: Boolean,
  val reconnect: Boolean,
)

data class AlertDelivery(
  val phone: Boolean,
  val band: Boolean,
  val phoneUrgency: PhoneUrgency?,
  val bandBehavior: BandAlertBehavior?,
) {
  companion object {
    val None =
      AlertDelivery(
        phone = false,
        band = false,
        phoneUrgency = null,
        bandBehavior = null,
      )
  }
}

class NotificationPolicy private constructor(
  private val mode: NotificationMode,
  private val waitingForReview: Boolean,
  private val needsAuthorization: Boolean,
  private val phone: Boolean,
  private val band: Boolean,
) {
  fun plan(state: TaskState, context: AlertContext): AlertDelivery {
    val eventEnabled =
      when (state) {
        TaskState.Running -> false
        TaskState.NeedsAuthorization -> needsAuthorization
        TaskState.WaitingForReview -> waitingForReview
      }
    val timingAllows =
      when (mode) {
        NotificationMode.Never -> false
        NotificationMode.WhenChatGptUnfocused -> !context.chatGptFocused
        NotificationMode.Always -> true
      }
    if (
      !eventEnabled ||
        !timingAllows ||
        context.androidForeground ||
        (context.reconnect && state == TaskState.WaitingForReview) ||
        (!phone && !band)
    ) {
      return AlertDelivery.None
    }

    return AlertDelivery(
      phone = phone,
      band = band,
      phoneUrgency =
        if (phone) {
          when (state) {
            TaskState.NeedsAuthorization -> PhoneUrgency.Vibrate
            TaskState.WaitingForReview -> PhoneUrgency.Silent
            TaskState.Running -> null
          }
        } else {
          null
        },
      bandBehavior = if (band) BandAlertBehavior.SystemControlled else null,
    )
  }

  companion object {
    fun create(
      mode: NotificationMode,
      waitingForReview: Boolean,
      needsAuthorization: Boolean,
      phone: Boolean,
      band: Boolean,
    ) =
      NotificationPolicy(
        mode = mode,
        waitingForReview = waitingForReview,
        needsAuthorization = needsAuthorization,
        phone = phone,
        band = band,
      )

    fun default() =
      create(
        mode = NotificationMode.WhenChatGptUnfocused,
        waitingForReview = true,
        needsAuthorization = true,
        phone = true,
        band = true,
      )
  }
}

File: windows-native/tests/notification_policy.rs
use codex_quota_windows_core::{
    AlertContext, AlertDelivery, AlertUrgency, NotificationMode, NotificationPolicy, TaskState,
};

#[test]
fn default_policy_only_alerts_after_chatgpt_loses_focus() {
    let policy = NotificationPolicy::default();

    assert_eq!(
        policy.plan(
            TaskState::WaitingForReview,
            AlertContext {
                chatgpt_focused: true,
                android_foreground: false,
                reconnect: false,
            },
        ),
        AlertDelivery::none()
    );
    assert_eq!(
        policy.plan(
            TaskState::Running,
            AlertContext {
                chatgpt_focused: false,
                android_foreground: false,
                reconnect: false,
            },
        ),
        AlertDelivery::none()
    );
    assert_eq!(
        policy.plan(
            TaskState::WaitingForReview,
            AlertContext {
                chatgpt_focused: false,
                android_foreground: false,
                reconnect: false,
            },
        ),
        AlertDelivery {
            phone: true,
            band: true,
            urgency: Some(AlertUrgency::Silent),
        }
    );
}

#[test]
fn android_foreground_suppresses_even_always_mode() {
    let policy = NotificationPolicy::new(NotificationMode::Always, true, true, true, true);

    assert_eq!(
        policy.plan(
            TaskState::NeedsAuthorization,
            AlertContext {
                chatgpt_focused: false,
                android_foreground: true,
                reconnect: false,
            },
        ),
        AlertDelivery::none()
    );
}

#[test]
fn reconnect_only_alerts_for_unresolved_authorization() {
    let policy = NotificationPolicy::default();
    let reconnect = AlertContext {
        chatgpt_focused: false,
        android_foreground: false,
        reconnect: true,
    };

    assert_eq!(
        policy.plan(TaskState::WaitingForReview, reconnect),
        AlertDelivery::none()
    );
    assert_eq!(
        policy.plan(TaskState::NeedsAuthorization, reconnect),
        AlertDelivery {
            phone: true,
            band: true,
            urgency: Some(AlertUrgency::Vibrate),
        }
    );
}

#[test]
fn event_and_channel_switches_are_independent() {
    let policy = NotificationPolicy::new(NotificationMode::Always, false, true, false, true);
    let background = AlertContext {
        chatgpt_focused: true,
        android_foreground: false,
        reconnect: false,
    };

    assert_eq!(
        policy.plan(TaskState::WaitingForReview, background),
        AlertDelivery::none()
    );
    assert_eq!(
        policy.plan(TaskState::NeedsAuthorization, background),
        AlertDelivery {
            phone: false,
            band: true,
            urgency: Some(AlertUrgency::Vibrate),
        }
    );
}
